### PR TITLE
KAFKA-9156: fix LazyTimeIndex & LazyOffsetIndex concurrency

### PR DESCRIPTION
**Pull request by alexandrfox**
Fix of https://issues.apache.org/jira/browse/KAFKA-9156

Race condition in concurrent  `get` method invocation of lazy indexes might lead to multiple `OffsetIndex`/`TimeIndex` objects being concurrently created. When that happens position of `MappedByteBuffer` in `AbstractIndex` is advanced to the last entry which in turn leads to a critical `BufferOverflowException` being thrown whenever leader or replica tries to append a record to the segment.

Moreover, `file_=` setter is seemingly also vulnerable to the race, since multiple threads can attempt to set a new file reference as well as create new Time/OffsetIndex objects at the same time. This might lead to the discrepant File references being held by e.g. LazyTimeIndex and its TimeIndex counterpart.

This patch attempts to fix the issue by making sure that index objects are atomically constructed when loaded lazily via `get` method. Additionally, `file` reference modifications are also made atomic and thread safe.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/TimeIndex.scala
Comment (lushilin) [inline core/src/main/scala/kafka/log/TimeIndex.scala]: i think this lock is too big.this lock can move to timeIndex is not define, and can use double check

Comment (alexandrfox) [reply core/src/main/scala/kafka/log/TimeIndex.scala]: considering how infrequently this is called I am not sure that additional double-check locking complication is needed here, but I am open to modify it in case others agree as well

File: core/src/main/scala/kafka/log/TimeIndex.scala
Comment (lushilin) [inline core/src/main/scala/kafka/log/TimeIndex.scala]: i think this lock is useless because this is a get file method.if modify the file, also can get correct file at next time.

Comment (alexandrfox) [reply core/src/main/scala/kafka/log/TimeIndex.scala]: I believe getter is also susceptible for race. Consider the following situation:
 - Thread 1 enters "else" codepath of the `file` getter and is ready to return the reference of `_file`
 - At this very moment Thread 2 performs two operations consecutively: calls `get` and immediately `file_=` setter replacing the current `file` reference within `TimeIndex` (but leaving `LazyTimeIndex`'s `_file` outdated
 - Thread 1 continues its operation and returns a stale reference to the File object which is no longer correct

File: core/src/main/scala/kafka/log/OffsetIndex.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/OffsetIndex.scala]: The usual pattern for something like this is to read to a local var first. And avoid another volatile read for the common case.

File: core/src/main/scala/kafka/log/TimeIndex.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/TimeIndex.scala]: I think a better way to handle this is to replace `Option[TimeIndex]` with two case classes that inherit from a common sealed trait. Then we don't have this issue where we have to keep 2 separate fields in sync.

Comment (alexandrfox) [reply core/src/main/scala/kafka/log/TimeIndex.scala]: @ijuma agreed, please check out the latest commit

File: core/src/main/scala/kafka/log/OffsetIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/OffsetIndex.scala]: @ijuma good catch, fixed

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [inline core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: unfortunately we cannot simply move this setter logic into the trait/case classes as well, since then we'd need to declare `indexFile` as `@volatile var indexFile` in `UninitializedLazyIndex` case class. This won't work for scala 2.11 due to https://github.com/scala/bug/issues/8873

Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: We don't need them to be case classes though. A plain old class will do. Also, we should move the sealed trait and subclasses to a companion object since they don't seem to need access to the outer class.

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: With the match based approach, we don't actually need this outer local `var` since the pattern match introduces a local val.

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: Should this be `protected`?

**Comment by ijuma** [comment]
Thanks for the update, looking better. I left a couple more comments.

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: Also, should this be `createIndex` or `newIndex`? `initializeIndex` sounds like it's taking an index and initializing it.

File: core/src/main/scala/kafka/log/OffsetIndex.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/OffsetIndex.scala]: We can remove `private val` from `_file` here and in `LazyTimeIndex`. That's the default for classes.

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: ++ for protected. I called it `createIndex` initially but felt a bit weird since it doesn't actually create the index file, rather just initialising a wrapper around it.

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: good to know!

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: That's a fair point. We should add some scaladoc. Maybe `loadIndex`?

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: We should probably name this differently since it's doing a lot more work than just assigning the new file.

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: yeah, I like `loadIndex` as well, — that makes more sense since it complements the "lazy" part, will add the javadoc

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [inline core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: I am wondering if we should keep this in its own file, or move-in with `AbstractIndex.scala` since both `LazyTimeIndex` and `LazyOffsetIndex` are kept together with `TimeIndex` and `OffsetIndex` definitions. Alternatively we can move them out into their own files. WDYT @ijuma?

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: this is now simplified since file modification logic was moved into the trait subclasses (I dropped `case class` so now it's possible to use `@volatile` again in scala 2.11), — I believe name still makes sense, how do you think?

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: agreed, moved into companion object

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: Makes sense to keep it in the same file as the `Lazy*Index` implementations.

Comment (ijuma) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: Oh, I see. The current structure is a bit weird. I think I'd keep all the Lazy implementations together. If you don't mind, I'll push a commit to your branch and we can keep/remove/adjust after.

**Comment by ijuma** [comment]
For some reason, I cannot push to your branch, so I submitted a PR: https://github.com/bookingcom/kafka/pull/1/files

File: core/src/main/scala/kafka/log/AbstractLazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/AbstractLazyIndex.scala]: sounds good!

**Comment by alexandrfox** [comment]
> For some reason, I cannot push to your branch, so I submitted a PR: https://github.com/bookingcom/kafka/pull/1/files

Probably due to org. settings. LGTM, so I merged it, thank you! Also, added a few missing type annotations to `_maxEntries` and `_entries` of `AbsrtactIndex`

File: core/src/main/scala/kafka/log/LazyIndex.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/LazyIndex.scala:39]: We should mention that the methods in this class are thread-safe, but to check the underlying `AbstractIndex` subclass documentation for information on their thread safety.

**Review by ijuma — APPROVED** [review]
LGTM, just a minor documentation suggestion below. It would be good to get a second opinion from @junrao or @hachikuji before we merge this to 2.3, 2.4 and trunk.

**Comment by ijuma** [comment]
@alexandrfox In the PR description, it's worth clarifying that mutation operations are done under a lock in the caller classes. I think the gap is that `get` can be called during `Log.read` which does not acquire any lock.

**Comment by alexandrfox** [comment]
> @alexandrfox In the PR description, it's worth clarifying that mutation operations are done under a lock in the caller classes. I think the gap is that `get` can be called during `Log.read` which does not acquire any lock.

Makes sense, I think LogCleaner also may call `get` while performing `doClean` (which can naturally happen at the same time as `read`).

File: core/src/main/scala/kafka/log/LazyIndex.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/LazyIndex.scala:82]: Does file need to be var? It seems that it's never updated.

**Review by junrao — APPROVED** [review]
@alexandrfox : Thanks for the PR. LGTM. Just a minor comment below.

**Review by hachikuji — APPROVED** [review]
LGTM. Nice fix and thanks for the additional cleanup!

File: core/src/main/scala/kafka/log/LazyIndex.scala
Comment (alexandrfox) [reply core/src/main/scala/kafka/log/LazyIndex.scala:82]: hey @junrao, yes, this reference is updated when `LazyIndex.file_=` setter is invoked before `LazyIndex.get`

**Comment by alexandrfox** [comment]
@ijuma looks like we have approvals, thanks to @junrao and @hachikuji. And thank you for taking your time guiding me through my first PR!

Also, thanks to @lushilin for pin-pointing the problem!

File: core/src/main/scala/kafka/log/LazyIndex.scala
Comment (junrao) [reply core/src/main/scala/kafka/log/LazyIndex.scala:82]: @alexandrfox : Thanks. That makes sense. So, we can keep the code as it is.

**Comment by ijuma** [comment]
One job passed, two jobs failed with known flaky tests:

> kafka.admin.ReassignPartitionsClusterTest.shouldListMovingPartitionsThroughApi
> kafka.api.ConsumerBounceTest.testClose
> kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup

**Comment by ijuma** [comment]
Merged to trunk, 2.4 and 2.3.

**Pull request merged**
